### PR TITLE
[LSP] Add `RenameRequest` and `PrepareRenameRequest`

### PR DESCRIPTION
**Pull request by Snowy1803**
Add support for renaming in LanguageServerProtocol. Rename options are restricted to pre-3.16 options, the ones added in 3.16, such as prepareSupportDefaultBehavior, aren't supported.

### CODE REVIEW

File: Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  /// The document location at which to lookup symbol information.
+  public var position: Position
+
+  public init(textDocument: TextDocumentIdentifier, position: Position) {
+    self.textDocument = textDocument
+    self.position = position
+  }
+}
+
+public struct PrepareRenameResponse: ResponseType, Hashable {
```
Comment (DavidGoldman) [inline Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:41]: Looks like in the spec it's defined as `Range | { range: Range, placeholder: string } | { defaultBehavior: boolean } | null` - do we need to handle the defaultBehavior case too?

**Comment by DavidGoldman** [comment]
@swift-ci please test

File: Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift (hunk 1 above)
Comment (Snowy1803) [reply Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:41]: defaultBehavior was added in LSP 3.16, and as I said, I’m not supporting that (would require adding it to the client capability, adding the appropriate rename option for what the default behavior is, etc). IMO it isn’t necessary

File: Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// Request to test the validity of a rename operation at the given location.
+///
+/// Looks up the symbol at the given position and returns the range of the symbol the user is renaming,
+/// or null if there is no symbol to be renamed at the given position
+///
+/// Servers that provide rename preparation should set`prepareProvider` to true in the `renameProvider` server capability.
```
Comment (ahoppen) [inline Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:18]: Missing space 
```suggestion
/// Servers that provide rename preparation should set `prepareProvider` to true in the `renameProvider` server capability.
```

File: Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:41]: I think it’s interesting that LSP added a new response option with a new version of LSP. AFAICT the server can’t know whether the client supports this response option or am I missing something?

---

In any case, wouldn’t it be cleaner to explicitly model the three different response types by e.g. making `PrepareRenameResponse` as an `enum`? As it stands now, we use `placeholder == nil` as a flag to convey whether the response was a raw `Range` or `{ range: Range, placeholder: string }`. This means that 
```json
{
  "range": …,
  "placeholder": null
}
```
wouldn’t round-trip. I know, the spec requires `placeholder` to be non-optional but still, we might as well prepare for `defaultBehavior` now.

**Comment by ahoppen** [comment]
Thanks for contributing to SourceKit-LSP @Snowy1803. I added one more response to @DavidGoldman’s comment, otherwise it’s looking good to me 👍.

File: Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift (hunk 1 above)
Comment (Snowy1803) [reply Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:41]: Client Capabilities have an additional `renameProvider.prepareSupportDefaultBehavior` property to specify support

Comment (ahoppen) [reply Sources/LanguageServerProtocol/Requests/PrepareRenameRequest.swift:41]: Ah, that makes sense. I’d be fine to merge this as-is.

**Review by ahoppen — APPROVED** [review]

**Comment by Snowy1803** [comment]
@DavidGoldman @benlangmuir May I ask what is needed in order to get this PR merged?

**Comment by ahoppen** [comment]
Thanks for the ping @Snowy1803, I didn’t have the PR on my radar anymore.

I think the comment https://github.com/apple/sourcekit-lsp/pull/431#discussion_r721365241 is still open. Could you comment on that thread?

**Comment by Snowy1803** [comment]
It has already been answered, hasn't it? As long as the client capability `renameProvider.prepareSupportDefaultBehavior` isn't provided, we don't have to support `defaultBehavior`

**Comment by ahoppen** [comment]
I’m sorry @Snowy1803. I was looking at the wrong view in GitHub that didn’t show your reply.

**Comment by ahoppen** [comment]
@swift-ci Please test and merge

**Comment by Snowy1803** [comment]
Do we have to wait some more or is there an issue with the bot?

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
Forgot that this repo doesn’t support testing + merging in one step.

**Comment by Snowy1803** [comment]
That's odd

**Pull request merged**